### PR TITLE
XDD-183: Cosmos results as json

### PR DESCRIPTION
**Pull request by mwestphall**
- Add the SKEMA parquet-to-json utility script
- Create a set of endpoints that extract a given parquet file from a cached result zip file, then return it converted to JSON
- Create an endpoint to extract an image from a cached zip file and return it as an image
- Add a check for whether a PDF is already present in cache before re-processing
- Update Jupyter notebook with new endpoints
- Refactor directory structure a bit since the application will be doing more things

### CODE REVIEW

File: cosmos_service/src/app.py [hunk 1]
```diff
@@ -109,15 +132,39 @@ def get_processing_result(job_id: str):
...
+@app.get("/process/{job_id}/result/text")
+def get_processing_result_text_segments(job_id: str, request: Request):
+    """
+    Return the text segments extracted by COSMOS and their bounding boxes as a list of JSON objects
+    """
+    job = get_job_details(job_id)
+    return convert_parquet_to_json(job, f'{job.pdf_name}.parquet', request)
+
+@app.get("/process/{job_id}/result/extractions/{extraction_type}")
+def get_processing_result_text_segments(job_id: str, extraction_type: str, request: Request):
```
Comment (iross) [inline cosmos_service/src/app.py:148]: Rename this

**Review by iross — COMMENTED** [review]
Looks good! A few questions above. I think the biggest change I'd like to see is the inclusion of these JSON responses within the zipped response (with image paths referring to the local versions. I really like the ability to do the lookup directly within the service, but I suspect TA4 might want to have a locally referenced JSON representation.

A few other thoughts:

- The `img_pth` change to point to craft the web URL is nice, but I'm slightly concerned about the implication of its persistence. I don't think it's a huge deal as long as the documentation is very clear about the retention policies.
- For the page-level images (which are the `img_pth` for the text/section returns), we do cleanup by default, and so those URLs has a 404 response.
- Did your equation segmentation from the hackathon make it into the make_parquet script?

**Comment by mwestphall** [comment]
@iross re: those questions

- The fact that the images disappear after the cache retention period is over is definitely an issue. I think long term storage of anything is outside the scope of this API, so we'll definitely need to document things well
- The parquet-to-json script should be updated to filter out the img_pth value from the text segment now
- The JSON data should now be included in the full zip response as well
- The changes to equation segmentation live in `ingest.process.aggregation.aggregate`, which is installed as a python package in the docker layer below this one (`uw-cosmos/cosmos-ingestion`). I believe that image is updated in dockerhub and this image is built on top of that updated version

**Comment by iross** [comment]
Looks good! Merging.

**Pull request merged**